### PR TITLE
Use the `babel-plugin-polyfill-*` packages in `preset-env`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #11823
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR is now targeting `main` without any feature flags because now the polyfill packages are transpiled to be compatible with Node.js 6, so they don't introduce breaking changes.

We will still need to update `transform-runtime`, but that is harder because the `babel-plugin-polyfill-*` packages don't inject imports to `@babel/runtime`.

This PR depends on https://github.com/babel/babel-polyfills/pull/50, https://github.com/babel/babel-polyfills/pull/51 and babel/babel-polyfills#52, ~~so I expect CI to fail for now~~. (**EDIT**: These PRs have been merged and published)

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12583"></a>

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12583"></a>

### CODE REVIEW

File: packages/babel-preset-env/test/fixtures/corejs2/usage-browserslist-config-ignore/stdout.txt
Comment (nicolo-ribaudo) [inline packages/babel-preset-env/test/fixtures/corejs2/usage-browserslist-config-ignore/stdout.txt:50]: The targets are now printed multiple times: once by preset-env (for the transform) and one by the polyfill plugin. I could inject a secret parameter used only by `preset-env` to avoid printing them in the polyfill provider, but I don't think that it's worth the complexity.

File: packages/babel-preset-env/test/fixtures/corejs2/usage-evaluated-not-confident/output.mjs
Comment (nicolo-ribaudo) [inline packages/babel-preset-env/test/fixtures/corejs2/usage-evaluated-not-confident/output.mjs:2]: I'm not sure because the old code included these polyfills, but the new one seems correct to me.

File: packages/babel-preset-env/package.json
Comment (merceyz) [inline packages/babel-preset-env/package.json:83]: Due to [how @babel/helper-define-polyfill-provider is implemented](https://github.com/babel/babel-polyfills/blob/5d2cea5ed86b4ccef4c54b03fd9a3c2791a578d2/packages/babel-helper-define-polyfill-provider/src/dependencies.js#L35) these will use `resolve` to locate modules, could they be updated to use `require.resolve` by applying the plugin you made in https://github.com/babel/babel/pull/12439 to them?

File: packages/babel-plugin-proposal-async-generator-functions/package.json
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-async-generator-functions/package.json:30]: @merceyz If I don't update this to `0.0.11`, I get this error:
```
➤ YN0001: │ Error: Assertion failed: /home/nicolo/Documenti/dev/babel/babel/packages/babel-plugin-proposal-async-generator-functions/node_modules/@babel/types cannot merge nodes with different link types
```

Is it a Yarn bug?

Comment (larixer) [reply packages/babel-plugin-proposal-async-generator-functions/package.json:30]: @nicolo-ribaudo Yes, looks like a Yarn bug, I will take a look and report back

Comment (merceyz) [reply packages/babel-plugin-proposal-async-generator-functions/package.json:30]: Thanks for the ping, I forwarded it to @larixer as he has a better grip on this part of the codebase

---
Apparently I didn't send this comment earlier

File: packages/babel-preset-env/test/fixtures/corejs2/usage-shippedProposals/output.js
Comment (JLHwung) [inline packages/babel-preset-env/test/fixtures/corejs2/usage-shippedProposals/output.js:35]: Duplicate polyfill imports should have been avoided.

File: packages/babel-preset-env/test/fixtures/corejs2/usage-symbol-iterator-in/output.mjs
Comment (JLHwung) [inline packages/babel-preset-env/test/fixtures/corejs2/usage-symbol-iterator-in/output.mjs:3]: The original output seems better. I am not sure why array iterator should be pulled here.

Comment (nicolo-ribaudo) [reply packages/babel-preset-env/test/fixtures/corejs2/usage-symbol-iterator-in/output.mjs:3]: It's because `Symbol.iterator in arr` should return `true` if `arr` is an array, and this can only happen if we load the `core-js/modules/es6.array.iterator.js` polyfill.

Our core-js 3 logic was already smart in this case (I reused it for core-js 2), and it already loads that polyfill on `main`: https://babeljs.io/repl#?browsers=defaults&build=&builtIns=usage&spec=false&loose=false&code_lz=MoTwtgRg9gNgdASwC4FMBOBDJU0AIEB2uGaaA3EA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.12.12&externalPlugins=

File: packages/babel-preset-env/test/fixtures/corejs2/usage-shippedProposals/output.js
Comment (nicolo-ribaudo) [reply packages/babel-preset-env/test/fixtures/corejs2/usage-shippedProposals/output.js:35]: Fixed by https://github.com/babel/babel-polyfills/pull/55

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-proposal-async-generator-functions/package.json
Comment (larixer) [reply packages/babel-plugin-proposal-async-generator-functions/package.json]: @nicolo-ribaudo Is this line really needed:
https://github.com/nicolo-ribaudo/babel/blob/polyfill-providers/package.json#L82
I'm afraid with this line you have atypical dependency graph, when there are workspace nodes not only at the beginning of the graph, but deep inside graph and that might make the graph unrepresentable in node_modules layout. I'm still investigating and I'm not fully understand what is happening yet, but the graph you have seems very difficult with this line.

Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-async-generator-functions/package.json]: I think it's needed because:
- `@babel/parser` doesn't have a dependency on `@babel/types`, which is used for TS typings (before adding it we need to add/fix different TS types across the monorepo)
- To avoid failures, yarn uses `packageExtensions` to add a `@babel/types` dependency
- It pulls in `@babel/types` from npm, causing problems due to nominal types

If the problem is that line, we will fix it at some point and for now we'll make sure to have the same `babel-plugin-polyfill-corejs3` in all the packages to help yarn deduping (it's working with this exact setup).

Comment (larixer) [reply packages/babel-plugin-proposal-async-generator-functions/package.json]: @nicolo-ribaudo Thanks, for now I think its a Yarn bug which should be fixed, I'll let you know when I have more information.

Comment (larixer) [reply packages/babel-plugin-proposal-async-generator-functions/package.json]: @nicolo-ribaudo Looks like the bug happens due to hoisting result is not terminal, if another pass of hoisting is done for this repo the conflict on `@babel/types` version is gone. I have implemented multi-pass hoisting which finishes when the result is the same on two consecutive hoisting passes here:
https://github.com/yarnpkg/berry/pull/2394
For this repo 3 passes are needed, on the 1st and 2nd pass the hoisting layout changes and on 3rd pass nothing is changed, but it is still needed for algorithm to understand that we have a terminal result.
The downside of this approach is that it is  obviously slower, in `master` hoisting takes 380ms for me, and multi-pass hoisting takes `1.28s`. I'm still looking into ways to optimise multi-pass hoisting speed.

**Comment by nicolo-ribaudo** [comment]
This PR is marked as `Internal`, but I'd like to land it in a minor release because I'm not really sure that it doesn't include improvements.

File: packages/babel-preset-env/test/fixtures/corejs2/usage-browserslist-config-ignore/stdout.txt
Comment (existentialism) [reply packages/babel-preset-env/test/fixtures/corejs2/usage-browserslist-config-ignore/stdout.txt:50]: FWIW, I do think it's worth the complexity :)

**Review by existentialism — APPROVED** [review]

**Comment by kreig303** [comment]
`@babel/preset-env@v7.13.0` and up is breaking our `nuxt` apps. not sure what has changed. currently diagnosing...

**Comment by nicolo-ribaudo** [comment]
Maybe it's fixed by https://github.com/nuxt/nuxt.js/pull/8883?

**Comment by kreig303** [comment]
@nicolo-ribaudo that's the one! guess we'll just wait on the release. 🎉

**Comment by nicolo-ribaudo** [comment]
Actually, I think it has been released in nuxt v2.15.2 :grimacing:

**Comment by kreig303** [comment]
@nicolo-ribaudo noted, but we are still on the `2.14.x` series. i have pinned it in the repos from now until we move up. thanks!

**Pull request merged**
